Add charm architecture documentation (#49)
canonical/paas-charm: docs/explanation/charm-architecture.rst

Reviewer: Potential missing sections in this document:
1. Juju events -- nothing too detailed, but it might be nice to list the observed events 
2. Charm code overview

Author: > Potential missing sections in this document:
> 
> 1. Juju events -- nothing too detailed, but it might be nice to list the observed events
> 2. Charm code overview

Updated

      }
      Container_Boundary(web_app_container, "Workload Container") {
         Component(workload, "Workload", "Web Application", "Observes events; serves web requests")
      }
   }
   Rel(charm_logic, workload, "Supervises<br>process")

OCI images
----------

We use `Rockcraft <https://canonical-rockcraft.readthedocs-hosted.com/en/latest/>`_ to build OCI Images for the web app charm. 

.. seealso::

   `How to publish your charm on Charmhub <https://juju.is/docs/sdk/publishing>`_
   
   `Build a 12-factor app rock <https://documentation.ubuntu.com/rockcraft/en/latest/how-to/build-a-12-factor-app-rock/>`_

Metrics
-------
The provided support for metrics and tracing depends on the enabled extension.

.. seealso:: 

   `Charmcraft reference | Extensions <https://canonical-charmcraft.readthedocs-hosted.com/en/stable/reference/extensions/>`_.

Integrations
------------
The available integrations, including those that are already pre-populated, varies between extensions.

.. seealso::

   `Charmcraft reference | Extensions <https://canonical-charmcraft.readthedocs-hosted.com/en/stable/reference/extensions/>`_.

> Events
> ------

Reviewer: ```suggestion
Juju events
-----------
```

> Accordingly to the `Juju SDK <https://juju.is/docs/sdk/event>`_: "an event is a data structure that encapsulates part of the execution context of a charm".

Reviewer: I would remove this line entirely

Author: Done

> For the web charms, the following events are observed:

Reviewer: ```suggestion
For a web app charm, the following events are observed:
```

> 1. `\<container name\>_pebble_ready <https://juju.is/docs/sdk/container-name-pebble-ready-event>`_: fired on Kubernetes charms when the requested container is ready.

Reviewer: ```suggestion
1. `\<container name\>_pebble_ready <https://canonical-juju.readthedocs-hosted.com/en/3.6/user/reference/hook/#container-pebble-ready>`_: fired on Kubernetes charms when the requested container is ready.
```

> Action: validate the charm configuration, run pending migrations and restart the workload.

Reviewer: All the `Action:` lines don't start on a new line. I would either make the term bold like `**Action**:` or I would remove `Action:` entirely

23. `tracing_endpoint_removed <https://github.com/canonical/tempo-coordinator-k8s-operator>`_: fired when one of the receiver endpoints is removed.
Action: validate the charm configuration, run pending migrations and restart the workload.

24. `smtp_data_available <https://github.com/canonical/smtp-integrator-operator>`_: fired when new SMTP data is present in the relation.
Action: validate the charm configuration, run pending migrations and restart the workload.

25. `rotate_secret_key_action <https://canonical-juju.readthedocs-hosted.com/en/latest/user/reference/hook/#action-hooks>`_: fired when rotate-secret-key action is executed. 
Action: generate a new secret token for the application.

Charm code overview
-------------------

The :code:`src/paas_charm/charm.py` contains the charm logic that all supported frameworks will inherit and extend.
Each framework will define its entry point in its own :code:`charm.py` file, defining a class that will extend from :code:`PaasCharm`.

PaasCharm is the base class from which all Charms are formed, defined by `Ops  <https://juju.is/docs/sdk/ops>`_ (Python framework for developing charms).

.. seealso::

>    `Charm <https://juju.is/docs/sdk/constructs#heading--charm>`_

Reviewer: ```suggestion
   `Charm <https://canonical-juju.readthedocs-hosted.com/en/3.6/user/reference/charm/>`_
```

.. Copyright 2025 Canonical Ltd.
.. See LICENSE file for licensing details.
.. _charm-architecture:

Charm architecture
==================

> PaaS charm is a framework to easily deploy and operate your Flask or Django workloads and associated infrastructure, such

Reviewer: ```suggestion
PaaS charm is a framework to easily deploy and operate your Flask, Django, FastAPI or Go workloads and associated infrastructure, such
```
Assuming that it works the same way for FastAPI and Go

as databases and ingress, using open source tooling.

The resulting charm design leverages the `sidecar <https://kubernetes.io/blog/2015/06/the-distributed-system-toolkit-patterns/#example-1-sidecar-containers>`_ pattern to allow multiple containers in each pod with `Pebble <https://juju.is/docs/sdk/pebble>`_ running as the workload container’s entrypoint.

Charm architecture diagram
--------------------------

.. mermaid::

   C4Container
   System_Boundary(paas_charm, "PaaS Charm") {
      Container_Boundary(charm_container, "Charm Container") {
         Component(charm_logic, "Charm Logic", "Juju Operator Framework", "Controls application deployment & config")
      }
      Container_Boundary(paas_container, "Workload Container") {

>          Component(workload, "Workload", "Web Application", "Observes events, serves web requests")

Reviewer: ```suggestion
         Component(workload, "Workload", "Web Application", "Observes events; serves web requests")
```

Pebble `services` are configured through `layers <https://github.com/canonical/pebble#layer-specification>`_, and the following containers represent each one a layer forming the effective Pebble configuration, or `plan`:

1. An :code:`app` container, which contains the workload to run in any of the supported web frameworks.

As a result, if you run a :code:`kubectl get pods` on a namespace named for the Juju model you've deployed the PaaS charm into, you'll see something like the following:

.. code-block:: text

   NAME                          READY   STATUS    RESTARTS   AGE
   paas-0                        2/2     Running   0          6h4m

This shows there are 2 containers - the named above, as well as a container for the charm code itself.

And if you run :code:`kubectl describe pod paas-0`, all the containers will have as Command :code:`/charm/bin/pebble`. That's because Pebble is responsible for the processes startup as explained above.

OCI images
----------

We use `Rockcraft <https://canonical-rockcraft.readthedocs-hosted.com/en/latest/>`_ to build OCI Images for the PaaS charm. 

> > See more: `How to publish your charm on Charmhub <https://juju.is/docs/sdk/publishing>`_

Reviewer: ```suggestion
.. seealso::

   `How to publish your charm on Charmhub <https://juju.is/docs/sdk/publishing>`_
   
   `Build a 12-factor app rock <https://documentation.ubuntu.com/rockcraft/en/latest/how-to/build-a-12-factor-app-rock/>`_
```

Pebble is a lightweight, API-driven process supervisor that is responsible for configuring processes to run in a container and controlling those processes throughout the workload lifecycle.

Pebble `services` are configured through `layers <https://github.com/canonical/pebble#layer-specification>`_, and the following containers represent each one a layer forming the effective Pebble configuration, or `plan`:

1. An :code:`app` container, which contains the workload to run in any of the supported web frameworks.

As a result, if you run a :code:`kubectl get pods` on a namespace named for the Juju model you've deployed the web app charm into, you'll see something like the following:

.. code-block:: text

   NAME                          READY   STATUS    RESTARTS   AGE
   web-app-0                     2/2     Running   0          6h4m

This shows there are 2 containers - the named above, as well as a container for the charm code itself.

And if you run :code:`kubectl describe pod web-app-0`, all the containers will have as Command :code:`/charm/bin/pebble`. That's because Pebble is responsible for the processes startup as explained above.

OCI images
----------

> We use `Rockcraft <https://canonical-rockcraft.readthedocs-hosted.com/en/latest/>`_ to build OCI Images for the PaaS charm. 

Reviewer: ```suggestion
We use `Rockcraft <https://canonical-rockcraft.readthedocs-hosted.com/en/latest/>`_ to build OCI Images for the web app charm. 
```

.. seealso::

   `How to publish your charm on Charmhub <https://juju.is/docs/sdk/publishing>`_
   
   `Build a 12-factor app rock <https://documentation.ubuntu.com/rockcraft/en/latest/how-to/build-a-12-factor-app-rock/>`_

Metrics
-------

> Depending on the enabled extension, different support for metrics and tracing is provided.

Reviewer: ```suggestion
The provided support for metrics and tracing depends on the enabled extension.
```

> See more: `extensions charmcraft documentation <https://canonical-charmcraft.readthedocs-hosted.com/en/stable/reference/extensions/>`_.

Integrations
------------

> The list of available integrations and those that are already pre-populated veries from extension to extensions.

Reviewer: ```suggestion
The available integrations, including those that are already pre-populated, varies between extensions.
```
